File: unity-package/Editor/Tools/ResEdit/EditTexture.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;
using UnityMcp.Models; // For Response class

namespace UnityMcp.Tools
{
    /// <summary>
    /// Handles texture import settings and modification operations.
    /// 对应方法名: manage_texture
    /// </summary>
    [ToolName("edit_texture")]
    public class EditTexture : StateMethodBase
    {
        /// <summary>
        /// 创建当前方法支持的参数键列表
        /// </summary>
        protected override MethodKey[] CreateKeys()
        {
            return new[]
            {
                new MethodKey("action", "操作类型：set_type, set_sprite_settings, get_settings", false),
                new MethodKey("texture_path", "纹理资源路径", false),
                new MethodKey("texture_type", "纹理类型：Default, NormalMap, EditorGUIAndLegacy, Sprite, Cursor, Cookie, Lightmap, HDR", true),
                new MethodKey("sprite_mode", "Sprite模式：Single, Multiple, Polygon", true),
                new MethodKey("pixels_per_unit", "每单位像素数", true),
                new MethodKey("sprite_pivot", "Sprite轴心：Center, TopLeft, TopCenter, TopRight, MiddleLeft, MiddleCenter, MiddleRight, BottomLeft, BottomCenter, BottomRight, Custom", true),
                new MethodKey("generate_physics_shape", "生成物理形状", true),
                new MethodKey("mesh_type", "网格类型：FullRect, Tight", true),
                new MethodKey("extrude_edges", "边缘挤出", true),
                new MethodKey("compression", "压缩格式：Uncompressed, LowQuality, NormalQuality, HighQuality", true),
                new MethodKey("max_texture_size", "最大纹理尺寸：32, 64, 128, 256, 512, 1024, 2048, 4096, 8192", true),
                new MethodKey("filter_mode", "过滤模式：Point, Bilinear, Trilinear", true),
                new MethodKey("wrap_mode", "包装模式：Repeat, Clamp, Mirror, MirrorOnce", true),
                new MethodKey("readable", "可读写", true),
                new MethodKey("generate_mip_maps", "生成Mip贴图", true),
                new MethodKey("srgb_texture", "sRGB纹理", true)
            };
        }

        protected override StateTree CreateStateTree()
        {
            return StateTreeBuilder
                .Create()
                .Key("action")
                    .Leaf("set_type", HandleSetTypeAction)
                    .Leaf("set_sprite_settings", HandleSetSpriteSettingsAction)
                    .Leaf("get_settings", HandleGetSettingsAction)
                .Build();
        }

        // --- State Tree Action Handlers ---

        /// <summary>
        /// 处理设置纹理类型的操作
        /// </summary>
        private object HandleSetTypeAction(JObject args)
        {
            string texturePath = args["texture_path"]?.ToString();
            string textureType = args["texture_type"]?.ToString();

            if (string.IsNullOrEmpty(texturePath))
            {
                return Response.Error("'texture_path' parameter is required.");
            }

            if (string.IsNullOrEmpty(textureType))
            {
                return Response.Error("'texture_type' parameter is required.");
            }

            LogInfo($"[ManageTexture] Setting texture type to '{textureType}' for '{texturePath}'");
            return SetTextureType(texturePath, textureType);
        }

        /// <summary>
        /// 处理设置Sprite设置的操作
        /// </summary>
        private object HandleSetSpriteSettingsAction(JObject args)
        {
            string texturePath = args["texture_path"]?.ToString();

            if (string.IsNullOrEmpty(texturePath))
            {
                return Response.Error("'texture_path' parameter is required.");
            }

            LogInfo($"[ManageTexture] Setting sprite settings for '{texturePath}'");
            return SetSpriteSettings(args, texturePath);
        }

        /// <summary>
        /// 处理获取纹理设置的操作
        /// </summary>
        private object HandleGetSettingsAction(JObject args)
        {
            string texturePath = args["texture_path"]?.ToString();

            if (string.IsNullOrEmpty(texturePath))
            {
                return Response.Error("'texture_path' parameter is required.");
            }

            LogInfo($"[ManageTexture] Getting texture settings for '{texturePath}'");
            return GetTextureSettings(texturePath);
        }

        // --- Core Methods ---

        /// <summary>
        /// 设置纹理类型
        /// </summary>
        private object SetTextureType(string texturePath, string textureType)
        {
            try
            {
                // 获取纹理导入器
                TextureImporter textureImporter = GetTextureImporter(texturePath);
                if (textureImporter == null)
                {
                    return Response.Error($"Could not get TextureImporter for '{texturePath}'. Make sure the path is correct and points to a texture.");
                }

                // 解析纹理类型
                if (!Enum.TryParse<TextureImporterType>(textureType, true, out TextureImporterType type))
                {
                    return Response.Error($"Invalid texture type '{textureType}'. Valid types: {string.Join(", ", Enum.GetNames(typeof(TextureImporterType)))}");
                }

                // 设置纹理类型
                textureImporter.textureType = type;

                // 如果设置为Sprite，自动配置常用的Sprite设置
                if (type == TextureImporterType.Sprite)
                {
                    textureImporter.spriteImportMode = SpriteImportMode.Single;
                    textureImporter.spritePixelsPerUnit = 100f;
                    textureImporter.spritePivot = Vector2.one * 0.5f; // Center
                }

                // 应用设置并重新导入
                EditorUtility.SetDirty(textureImporter);
                textureImporter.SaveAndReimport();

                LogInfo($"[ManageTexture] Successfully set texture type to '{textureType}' for '{texturePath}'");
                return Response.Success($"Texture type set to '{textureType}' for '{texturePath}'.");
            }
            catch (Exception e)
            {
                LogInfo($"[ManageTexture] Error setting texture type: {e.Message}");
                return Response.Error($"Error setting texture type: {e.Message}");
            }
        }

        /// <summary>
        /// 设置Sprite设置
        /// </summary>
        private object SetSpriteSettings(JObject args, string texturePath)
        {
            try
            {
                // 获取纹理导入器
                TextureImporter textureImporter = GetTextureImporter(texturePath);
                if (textureImporter == null)
                {
                    return Response.Error($"Could not get TextureImporter for '{texturePath}'. Make sure the path is correct and points to a texture.");
                }

                // 确保是Sprite类型
                if (textureImporter.textureType != TextureImporterType.Sprite)
                {
                    textureImporter.textureType = TextureImporterType.Sprite;
                }

                // 设置Sprite模式
                string spriteMode = args["sprite_mode"]?.ToString();
                if (!string.IsNullOrEmpty(spriteMode))
                {
                    if (Enum.TryParse<SpriteImportMode>(spriteMode, true, out SpriteImportMode mode))
                    {
                        textureImporter.spriteImportMode = mode;
                    }
                }

                // 设置每单位像素数
                float pixelsPerUnit = args["pixels_per_unit"]?.ToObject<float>() ?? 100f;
                textureImporter.spritePixelsPerUnit = pixelsPerUnit;

                // 设置轴心点
                string spritePivot = args["sprite_pivot"]?.ToString();
                if (!string.IsNullOrEmpty(spritePivot))
                {
                    Vector2 pivot = GetPivotVector(spritePivot);
                    textureImporter.spritePivot = pivot;
                }

                // 注意：某些高级Sprite设置可能在不同Unity版本中不可用
                // 这些设置通常通过Unity Editor UI手动配置

                // 设置压缩和质量
                SetTextureCompressionSettings(textureImporter, args);

                // 设置其他通用设置
                SetGeneralTextureSettings(textureImporter, args);

                // 应用设置并重新导入
                EditorUtility.SetDirty(textureImporter);
                textureImporter.SaveAndReimport();

                LogInfo($"[ManageTexture] Successfully applied sprite settings to '{texturePath}'");
                return Response.Success($"Sprite settings applied to '{texturePath}'.");
            }
            catch (Exception e)
            {
                LogInfo($"[ManageTexture] Error setting sprite settings: {e.Message}");
                return Response.Error($"Error setting sprite settings: {e.Message}");
            }
        }

        /// <summary>
        /// 获取纹理设置
        /// </summary>
        private object GetTextureSettings(string texturePath)
        {
            try
            {
                TextureImporter textureImporter = GetTextureImporter(texturePath);
                if (textureImporter == null)
                {
                    return Response.Error($"Could not get TextureImporter for '{texturePath}'. Make sure the path is correct and points to a texture.");
                }

                var settings = new
                {
                    textureType = textureImporter.textureType.ToString(),
                    spriteImportMode = textureImporter.spriteImportMode.ToString(),
                    spritePixelsPerUnit = textureImporter.spritePixelsPerUnit,
                    spritePivot = textureImporter.spritePivot,
                    maxTextureSize = textureImporter.maxTextureSize,
                    textureCompression = textureImporter.textureCompression.ToString(),
                    filterMode = textureImporter.filterMode.ToString(),
                    wrapMode = textureImporter.wrapMode.ToString(),
                    isReadable = textureImporter.isReadable,
                    mipmapEnabled = textureImporter.mipmapEnabled,
                    sRGBTexture = textureImporter.sRGBTexture
                };

                return Response.Success($"Retrieved texture settings for '{texturePath}'.", settings);
            }
            catch (Exception e)
            {
                LogInfo($"[ManageTexture] Error getting texture settings: {e.Message}");
                return Response.Error($"Error getting texture settings: {e.Message}");
            }
        }

        // --- Helper Methods ---

        /// <summary>
        /// 获取纹理导入器
        /// </summary>
        private TextureImporter GetTextureImporter(string texturePath)
        {
            try
            {
                return AssetImporter.GetAtPath(texturePath) as TextureImporter;
            }
            catch (Exception e)
            {
                LogInfo($"[ManageTexture] Error getting TextureImporter: {e.Message}");
                return null;
            }
        }

        /// <summary>
        /// 获取轴心点向量
        /// </summary>
        private Vector2 GetPivotVector(string pivotName)
        {
            switch (pivotName.ToLower())
            {
                case "center":
                case "middlecenter":
                    return new Vector2(0.5f, 0.5f);
                case "topleft":
                    return new Vector2(0f, 1f);
                case "topcenter":
                    return new Vector2(0.5f, 1f);
                case "topright":
                    return new Vector2(1f, 1f);
                case "middleleft":
                    return new Vector2(0f, 0.5f);
                case "middleright":
                    return new Vector2(1f, 0.5f);
                case "bottomleft":
                    return new Vector2(0f, 0f);
                case "bottomcenter":
                    return new Vector2(0.5f, 0f);
                case "bottomright":
                    return new Vector2(1f, 0f);
                default:
                    return new Vector2(0.5f, 0.5f); // Default to center
            }
        }

        /// <summary>
        /// 设置纹理压缩设置
        /// </summary>
        private void SetTextureCompressionSettings(TextureImporter textureImporter, JObject args)
        {
            // 设置压缩格式
            string compression = args["compression"]?.ToString();
            if (!string.IsNullOrEmpty(compression))
            {
                switch (compression.ToLower())
                {
                    case "uncompressed":
                        textureImporter.textureCompression = TextureImporterCompression.Uncompressed;
                        break;
                    case "lowquality":
                        textureImporter.textureCompression = TextureImporterCompression.CompressedLQ;
                        break;
                    case "normalquality":
                        textureImporter.textureCompression = TextureImporterCompression.Compressed;
                        break;
                    case "highquality":
                        textureImporter.textureCompression = TextureImporterCompression.CompressedHQ;
                        break;
                }
            }

            // 设置最大纹理尺寸
            int maxSize = args["max_texture_size"]?.ToObject<int>() ?? 2048;
            textureImporter.maxTextureSize = maxSize;
        }

        /// <summary>
        /// 设置通用纹理设置
        /// </summary>
        private void SetGeneralTextureSettings(TextureImporter textureImporter, JObject args)
        {
            // 设置过滤模式
            string filterMode = args["filter_mode"]?.ToString();
            if (!string.IsNullOrEmpty(filterMode))
            {
                if (Enum.TryParse<FilterMode>(filterMode, true, out FilterMode mode))
                {
                    textureImporter.filterMode = mode;
                }
            }

            // 设置包装模式
            string wrapMode = args["wrap_mode"]?.ToString();
            if (!string.IsNullOrEmpty(wrapMode))
            {
                if (Enum.TryParse<TextureWrapMode>(wrapMode, true, out TextureWrapMode mode))
                {
                    textureImporter.wrapMode = mode;
                }
            }

            // 设置是否可读
            bool readable = args["readable"]?.ToObject<bool>() ?? false;
            textureImporter.isReadable = readable;

            // 设置是否生成Mip贴图
            bool generateMipMaps = args["generate_mip_maps"]?.ToObject<bool>() ?? true;
            textureImporter.mipmapEnabled = generateMipMaps;

            // 设置sRGB纹理
            bool srgbTexture = args["srgb_texture"]?.ToObject<bool>() ?? true;
            textureImporter.sRGBTexture = srgbTexture;
        }
    }
}

File: server/tools/edit_texture.py
"""
纹理导入设置修改工具
修改Unity中纹理资源的导入设置，包括设置为Sprite类型、调整压缩质量等
"""

from typing import Dict, Any, Optional
from pydantic import Field
from mcp.server.fastmcp import FastMCP, Context
from unity_connection import get_unity_connection


def register_edit_texture_tools(mcp: FastMCP):
    @mcp.tool("edit_texture")
    def edit_texture(
        ctx: Context,
        action: str = Field(
            ...,
            title="操作类型",
            description="操作类型",
            examples=["set_type", "set_sprite_settings", "get_settings"]
        ),
        texture_path: str = Field(
            ...,
            title="纹理资源路径",
            description="纹理资源路径（相对于Assets）",
            examples=["Assets/Pics/rabbit.jpg", "Assets/Textures/icon.png", "Assets/UI/button.png"]
        ),
        texture_type: Optional[str] = Field(
            None,
            title="纹理类型",
            description="纹理类型",
            examples=["Default", "NormalMap", "Sprite", "Cursor", "Cookie", "Lightmap", "HDR"]
        ),
        sprite_mode: Optional[str] = Field(
            None,
            title="Sprite模式",
            description="Sprite模式",
            examples=["Single", "Multiple", "Polygon"]
        ),
        pixels_per_unit: Optional[float] = Field(
            None,
            title="每单位像素数",
            description="每单位像素数",
            examples=[100, 200, 1]
        ),
        sprite_pivot: Optional[str] = Field(
            None,
            title="Sprite轴心点",
            description="Sprite轴心点",
            examples=["Center", "TopLeft", "TopCenter", "TopRight", "MiddleLeft", "MiddleCenter", "MiddleRight", "BottomLeft", "BottomCenter", "BottomRight"]
        ),
        generate_physics_shape: Optional[bool] = Field(
            None,
            title="生成物理形状",
            description="生成物理形状"
        ),
        mesh_type: Optional[str] = Field(
            None,
            title="网格类型",
            description="网格类型"
        ),
        compression: Optional[str] = Field(
            None,
            title="压缩格式",
            description="压缩格式",
            examples=["HighQuality", "NormalQuality", "LowQuality"]
        ),
        max_texture_size: Optional[int] = Field(
            None,
            title="最大纹理尺寸",
            description="最大纹理尺寸",
            examples=[1024, 2048, 4096]
        ),
        filter_mode: Optional[str] = Field(
            None,
            title="过滤模式",
            description="过滤模式",
            examples=["Point", "Bilinear", "Trilinear"]
        ),
        wrap_mode: Optional[str] = Field(
            None,
            title="包装模式",
            description="包装模式",
            examples=["Repeat", "Clamp", "Mirror", "MirrorOnce"]
        ),
        readable: Optional[bool] = Field(
            None,
            title="可读写",
            description="可读写"
        ),
        generate_mip_maps: Optional[bool] = Field(
            None,
            title="生成Mip贴图",
            description="生成Mip贴图"
        ),
        srgb_texture: Optional[bool] = Field(
            None,
            title="sRGB纹理",
            description="sRGB纹理"
        )
    ) -> Dict[str, Any]:
        """
        纹理导入设置修改工具
        
        支持的操作:
        - set_type: 设置纹理类型
        - set_sprite_settings: 设置Sprite详细参数
        - get_settings: 获取当前纹理设置
        """
        try:
            unity_conn = get_unity_connection()
            
            # 构建命令参数
            cmd = {
                "action": action,
                "texture_path": texture_path
            }
            
            # 添加可选参数
            optional_params = {
                "texture_type": texture_type,
                "sprite_mode": sprite_mode,
                "pixels_per_unit": pixels_per_unit,
                "sprite_pivot": sprite_pivot,
                "generate_physics_shape": generate_physics_shape,
                "mesh_type": mesh_type,
                "compression": compression,
                "max_texture_size": max_texture_size,
                "filter_mode": filter_mode,
                "wrap_mode": wrap_mode,
                "readable": readable,
                "generate_mip_maps": generate_mip_maps,
                "srgb_texture": srgb_texture
            }
            
            for key, value in optional_params.items():
                if value is not None:
                    cmd[key] = value
            
            # 发送命令到Unity
            result = unity_conn.send_command_with_retry("edit_texture", cmd)
            
            return {
                "success": True,
                "message": f"Texture operation '{action}' completed successfully",
                "data": result.get("data", {}),
                "error": None
            }
            
        except Exception as e:
            return {
                "success": False,
                "message": f"Texture operation '{action}' failed",
                "data": {},
                "error": str(e)
            }
